### PR TITLE
Retrieve information from a EngineeringModel #fixes 160

### PR DESCRIPTION
**Pull request by jaimeatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
The basic retrieve of information from a 10-25 Engineering Model.

### CODE REVIEW

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 1]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
         /// </summary>
         private void InitializeElements()
         {
-            if(this.SessionAnchor != null)
+            var iteration = this.SessionAnchor?.OpenIteration;
+
+            var elementUsages = this.SessionAnchor?.OpenIteration?.Element.SelectMany(ed => ed.ContainedElement).OrderBy(x=>x.Name).ToList();
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:118]: I think in the UI you need to implement for the user to select the Option and ignore the EUs that are excluded in that Option.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 2]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
+            
+            if(elementUsages is not null)
             {
-                var iteration = this.SessionAnchor?.OpenIteration;
+                foreach (var elementUsage in elementUsages)
+                {
+                    var parameter = elementUsage.ElementDefinition.Parameter.FirstOrDefault(x => x.ParameterType.Name == "Shape kind"
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:124]: Check on ShortName please.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 3]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
+                                          && x.ParameterType is EnumerationParameterType
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:125]: I think you need to specify the order of these booleans with `()`

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 4]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
+                    Primitive basicShape = null;
+                    if (parameter is not null)
+                    {
+                        string? shapekind = parameter?.ExtractActualValues(1).First();
+                        switch (shapekind?.ToLowerInvariant())
+                        {
+                            case "box": basicShape = new Cube(1, 1, 1); break;
+                        }
+                    }
+                    parameter = elementUsage.ElementDefinition.Parameter.FirstOrDefault(x => x.ParameterType.Name == "Position"
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:137]: Again use ShortName.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 5]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
-            if(this.SessionAnchor != null)
+            var iteration = this.SessionAnchor?.OpenIteration;
+
+            var elementUsages = this.SessionAnchor?.OpenIteration?.Element.SelectMany(ed => ed.ContainedElement).OrderBy(x=>x.Name).ToList();
+            
+            if(elementUsages is not null)
             {
-                var iteration = this.SessionAnchor?.OpenIteration;
+                foreach (var elementUsage in elementUsages)
+                {
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:123]: You can extract the below code into own method for clarity.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 6]
```diff
@@ -122,31 +113,47 @@ protected async override Task OnAfterRenderAsync(bool firstRender)
...
+                        }
+                    }
+                    parameter = elementUsage.ElementDefinition.Parameter.FirstOrDefault(x => x.ParameterType.Name == "Position"
+                                                          && x.ParameterType is CompoundParameterType);
 
-                if (elementUsages != null)
-                {
-                    foreach (var elementUsage in elementUsages)
+                    string[]? translations = parameter?.ExtractActualValues(3);
+                    if (basicShape is PositionablePrimitives positionablePrim && translations is not null)
```
Comment (alexatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:141]: These kind of things should be refacottred to be in own methods that can be Unit Tested.

**Review by alexatrhea — CHANGES_REQUESTED** [review]
Some comments provided, and please also don't forget to add unit tests.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs (hunk 1 above)
Comment (jaimeatrhea) [reply COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:118]: I think that is another issue. [https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues/166](url)

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 7]
```diff
@@ -1,193 +1,177 @@
...
+                {
+                    JSInterop.JsRuntime = JsRuntime;
+                }
+                else
+                {
+                    throw new JSException("JSRuntime can't be null");
+                }
+ 
+                Scene.InitCanvas(this.CanvasReference);
+                InitializeElements();
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:111]: please prefix with `this.` -> `this.InitializeElements();`

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 8]
```diff
@@ -1,193 +1,177 @@
...
+        /// </summary>
+        private void InitializeElements()
+        {
+            var elementUsages = this.SessionAnchor?.OpenIteration?.Element.SelectMany(ed => ed.ContainedElement).OrderBy(x=>x.Name).ToList();
+            
+            if(elementUsages is not null)
+            {
+                foreach (var elementUsage in elementUsages)
+                {
+                    if(ShapeFactory.TryGetPrimitiveFromElementUsageParameter(elementUsage, out Primitive basicShape))
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:128]: please prefix with `this.` -> `this.ShapeFactory...`

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 9]
```diff
@@ -1,193 +1,177 @@
...
+            Scene.AddPrimitive(xAxis, Color.Red);
+
+            Line yAxis = new Line(0, -size, 0, 0, size, 0);
+            Scene.AddPrimitive(yAxis, Color.Green);
+
+            Line zAxis = new Line(0, 0, -size, 0, 0, size);
+            Scene.AddPrimitive(zAxis, Color.Blue);
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:177]: please add new line at the end

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 10]
```diff
@@ -1,193 +1,177 @@
...
+//
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:23]: please add a new line between copyright header and namespace declaration

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 11]
```diff
@@ -1,193 +1,177 @@
...
+                }
+                else
+                {
+                    throw new JSException("JSRuntime can't be null");
+                }
+ 
+                Scene.InitCanvas(this.CanvasReference);
+                InitializeElements();
+
+                AddWorldAxes();
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:113]: please prefix with `this.` -> `this.AddWorldAxes();`

File: COMETwebapp/ParameterExtensions.cs [hunk 12]
```diff
@@ -1,67 +1,70 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp
```
Comment (samatrhea) [inline COMETwebapp/ParameterExtensions.cs:24]: please insert newline between header and namespace

File: COMETwebapp/ParameterExtensions.cs [hunk 13]
```diff
@@ -1,67 +1,70 @@
...
+{
+    using CDP4Common.EngineeringModelData;
+
+    /// <summary>
+    /// Class containing extension methods for the <see cref="Parameter"/> or <see cref="ParameterOverride"/>
+    /// </summary>
+    public static class ParameterExtensions
```
Comment (samatrhea) [inline COMETwebapp/ParameterExtensions.cs:31]: this method should be part of the COMET-SDK, please add it there and remove it from this project once you are happy with its implementation

File: COMETwebapp/Primitives/Cone.cs [hunk 14]
```diff
@@ -1,74 +1,74 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cone.cs:24]: newline

File: COMETwebapp/Primitives/Cone.cs [hunk 15]
```diff
@@ -1,74 +1,74 @@
...
+        public Cone(double x, double y, double z, double radius, double height)
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Radius = radius;
+            this.Height = height;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cone.cs:74]: newline at end please

File: COMETwebapp/ParameterExtensions.cs [hunk 16]
```diff
@@ -1,67 +1,70 @@
...
+    {
+        /// <summary>
+        /// Extract the specified number of values from the parameter
+        /// </summary>
+        /// <param name="parameter">the <see cref="Parameter"/> to extract the data from</param>
+        /// <param name="numberOfValues">number of values to extract</param>
+        /// <returns>the extracted values</returns>
+        /// <exception cref="ArgumentException">if the parameter is null</exception>
+        public static string[] ExtractActualValues(this Parameter parameter, int numberOfValues)
+        {
```
Comment (samatrhea) [inline COMETwebapp/ParameterExtensions.cs:41]: please put argument checking at the beginning:
```
if (parameter == null)
{
    throw new ArgumentException("The parameter may not be null");
}
```

would be curious to understand how you determine the `numberOfValues` argument.

File: COMETwebapp/Primitives/Cube.cs [hunk 17]
```diff
@@ -1,83 +1,83 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cube.cs:24]: new line

File: COMETwebapp/Primitives/Cube.cs [hunk 18]
```diff
@@ -1,83 +1,83 @@
...
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Width = width;
+            this.Height = height;
+            this.Depth = depth;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cube.cs:83]: new line at end

File: COMETwebapp/Primitives/Cylinder.cs [hunk 19]
```diff
@@ -1,74 +1,74 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cylinder.cs:24]: new line

File: COMETwebapp/Primitives/Cylinder.cs [hunk 20]
```diff
@@ -1,74 +1,74 @@
...
+        public Cylinder(double x, double y, double z, double radius, double height)
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Radius = radius;
+            this.Height = height;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cylinder.cs:74]: new line

File: COMETwebapp/Primitives/IShapeFactory.cs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace COMETwebapp.Primitives
+{
+    using CDP4Common.EngineeringModelData;
+
+    public interface IShapeFactory
+    {
+        bool TryGetPrimitiveFromElementUsageParameter(ElementUsage elementUsage, out Primitive basicShape);
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/IShapeFactory.cs:33]: new line

File: COMETwebapp/Primitives/PositionablePrimitives.cs [hunk 22]
```diff
@@ -1,116 +1,152 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/PositionablePrimitives.cs:24]: new line

File: COMETwebapp/Primitives/PositionablePrimitives.cs [hunk 23]
```diff
@@ -1,116 +1,152 @@
...
+                {
+                    var x = double.Parse(translations[0]);
+                    var y = double.Parse(translations[1]);
+                    var z = double.Parse(translations[2]);
+                    this.SetTranslation(x, y, z);
+                }
+            }
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/PositionablePrimitives.cs:152]: new line

File: COMETwebapp/Primitives/ShapeFactory.cs [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace COMETwebapp.Primitives
+{
+    using CDP4Common.EngineeringModelData;
+    using CDP4Common.SiteDirectoryData;
+
+    public class ShapeFactory : IShapeFactory
+    {
+        /// <summary>
+        /// Tries to get a <see cref="Primitive"/> from the <see cref="ElementUsage"/>
```
Comment (samatrhea) [inline COMETwebapp/Primitives/ShapeFactory.cs:33]: `Tries to create a ...`

File: COMETwebapp/Primitives/ShapeFactory.cs [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+                }
+            }
+            else
+            {
+                basicShape = null;
+                return false;
+            }
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/ShapeFactory.cs:59]: new line

File: COMETwebapp/Primitives/Sphere.cs [hunk 26]
```diff
@@ -1,65 +1,65 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Sphere.cs:24]: new line

File: COMETwebapp/Primitives/Sphere.cs [hunk 27]
```diff
@@ -1,65 +1,65 @@
...
+        /// <param name="radius">the radius of the sphere</param>
+        public Sphere(double x, double y, double z, double radius)
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Radius = radius;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Sphere.cs:65]: new line

File: COMETwebapp/Primitives/Torus.cs [hunk 28]
```diff
@@ -1,74 +1,74 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Torus.cs:24]: new line

File: COMETwebapp/Primitives/Torus.cs [hunk 29]
```diff
@@ -1,74 +1,74 @@
...
+        public Torus(double x, double y, double z, double diameter, double thickness)
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Diameter = diameter;
+            this.Thickness = thickness;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Torus.cs:74]: new line

File: COMETwebapp/Program.cs [hunk 30]
```diff
@@ -1,62 +1,64 @@
...
+            builder.Services.AddSingleton<IVersionService, VersionService>();
+            builder.Services.AddSingleton<IShapeFactory, ShapeFactory>();
+
+            builder.Services.AddDevExpressBlazor();
+            builder.Services.AddBlazorStrap();
+
+            await builder.Build().RunAsync();
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Program.cs:64]: new line

**Review by samatrhea — CHANGES_REQUESTED** [review]

**Review by alexatrhea — APPROVED** [review]

**Review by samatrhea — APPROVED** [review]

**Pull request merged**